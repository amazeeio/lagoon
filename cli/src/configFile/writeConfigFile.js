// @flow

import yaml from 'js-yaml';
import R from 'ramda';
import { writeFile } from '../util/fs';

import { configFileInputOptionsTypes } from '../types/ConfigFile';
import type { ConfigFileInput } from '../types/ConfigFile';

export function writeConfigFile(
  filepath: string,
  inputOptions: ConfigFileInput,
): Promise<void> {
  const errors = [];

  const inputOptionsWithoutEmptyStrings = R.reduce(
    (acc, [optionKey, optionVal]) => {
      // Validate
      const optionType = R.prop(optionKey, configFileInputOptionsTypes);

      if (!optionType) {
        errors.push(
          `- Invalid config option "${optionKey}". Valid options: ${R.join(
            ', ',
            R.keys(configFileInputOptionsTypes),
          )}`,
        );
      } else if (!R.is(optionType, optionVal)) {
        errors.push(
          `- Invalid config option value for "${optionKey}": "${optionVal}" expected type: ${R.prop(
            'name',
            optionType,
          )}`,
        );
      }

      // Filter out empty strings
      if (!R.both(R.is(String), R.isEmpty)(optionVal)) {
        acc[optionKey] = optionVal;
      }

      return acc;
    },
    {},
    R.toPairs(inputOptions),
  );

  if (R.length(errors) > 0) throw new Error(errors.join('\n'));

  const yamlConfig = yaml.safeDump(inputOptionsWithoutEmptyStrings);
  return writeFile(filepath, yamlConfig);
}
