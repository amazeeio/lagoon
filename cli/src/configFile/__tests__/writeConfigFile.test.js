// @flow

import { writeConfigFile } from '../writeConfigFile';

jest.mock('../../util/fs');
const fs = require('../../util/fs');

// Flow does not know which objects are actual mocks
// this function casts given parameter to JestMockFn
const _castMockForFlow = (mockFn: any): JestMockFn<any, any> => mockFn;

describe('writeConfigFile', () => {
  it('should write default config to given path', async () => {
    writeConfigFile('lagoon.yml', {
      project: 'your_project',
      api: '',
      ssh: '',
      token: '',
    });
    const [filename, data] = _castMockForFlow(fs.writeFile).mock.calls[0];

    expect(filename).toBe('lagoon.yml');
    expect(data).toMatchSnapshot();
  });
});
