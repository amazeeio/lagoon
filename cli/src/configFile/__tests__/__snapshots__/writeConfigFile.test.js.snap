// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`writeConfigFile should write default config to given path 1`] = `
"project: your_project
"
`;
